casper/src/main/scala/coop/rchain/casper/engine/Running.scala

```diff
@@ -8,84 +8,217 @@ import cats.{Applicative, Monad}
 import com.google.protobuf.ByteString
 import coop.rchain.blockstorage.{BlockDagStorage, BlockStore}
 import coop.rchain.casper.LastApprovedBlock.LastApprovedBlock
-import coop.rchain.casper.MultiParentCasperRef.MultiParentCasperRef
 import coop.rchain.casper._
 import coop.rchain.casper.util.comm.CommUtil
+import coop.rchain.casper.util._
 import coop.rchain.casper.genesis.Genesis
 import coop.rchain.casper.protocol._
 import coop.rchain.casper.util.rholang.RuntimeManager
 import coop.rchain.catscontrib.Catscontrib._
 import coop.rchain.catscontrib.MonadTrans
 import coop.rchain.comm.discovery.NodeDiscovery
 import coop.rchain.comm.protocol.routing.Packet
+import coop.rchain.comm.rp.ProtocolHelper.packet
 import coop.rchain.comm.rp.Connect.{ConnectionsCell, RPConfAsk}
 import coop.rchain.comm.transport.{Blob, TransportLayer}
 import coop.rchain.comm.{transport, PeerNode}
 import coop.rchain.metrics.Metrics
 import coop.rchain.models.Validator.Validator
-import coop.rchain.shared.{Log, LogSource, Time}
+import coop.rchain.shared.{Cell, Log, LogSource, Time}
+import coop.rchain.catscontrib.Catscontrib, Catscontrib._
 import monix.eval.Task
 import monix.execution.Scheduler
-
-import scala.concurrent.duration.FiniteDuration
+import coop.rchain.models.BlockHash.BlockHash
+import scala.concurrent.duration._
 import scala.util.Try
 
 /** Node in this state has already received at least one [[ApprovedBlock]] and it has created an instance
   * of [[MultiParentCasper]].
   *
   * In the future it will be possible to create checkpoint with new [[ApprovedBlock]].
-    **/
-class Running[F[_]: RPConfAsk: BlockStore: Monad: ConnectionsCell: TransportLayer: Log: Time](
-    private val casper: MultiParentCasper[F],
+  **/
+object Running {
+
+  val timeout: FiniteDuration = 240 seconds
+
+  final case class Requested(
+      timestamp: Long,
+      peers: Set[PeerNode] = Set.empty,
```

Comment:
It seems `peers` are not used anywhere. You are moving peers from the waiting list to this set. But you could also just drop the peers.

Reply:
```

    "ignore if requested before for the same peer" in {
      // given
      val sender = peerNode("somePeer", 40400)
      val requestedBefore =
        Map(hash -> Requested(timestamp = System.currentTimeMillis, peers = Set(sender)))
      implicit val requestedBlocks = initRequestedBlocks(init = requestedBefore)
      val casperContains           = alwaysFalse
      // when
      Running.handleHasBlock[Coeval](sender, hb)(casperContains).apply()
      // then
      transport.requests shouldBe empty
    }
```